### ISSUE TITLE
[Feature] Add game data writers

### ISSUE THREAD

**Issue by Pyrdacor** [Pyrdacor/Ambermoon.net#53]
Those writers should allow to save game objects back to files. Useful for savegame writing, game data patching or modding.

Writers needed for:
- Party members
- Monsters
- NPCs
- Chests
- Merchants
- Automap
- Saves (savegame names)
- Savegame data (party data sav)
- Items
- Texts (NPC, party member, map, etc)
- Maps
    - Map data
    - Icon data
    - Lab data

Maybe also allow graphic writing later?

**Comment by a1exh** [comment]
How are the writers used?

**Comment by Pyrdacor** [comment]
At the moment you need to program in C#. So I can use them to change data. I need a game version which I should patch and a list of patches to perform.

**Comment by a1exh** [comment]
I'd like to start working on them on a file-by-file basis starting with v1.07, Party_char.amb.

If I am going too slow for you then for v1.07 take a look at the modified version by Meynaf

http://meynaf.free.fr/pr/amberpatch.lzx

The patch is not comprehensive (there are known fixable bugs he didn't fix and at least one with a mistake) and it contains fixes not detailed in the list of fixes in the text file included in the archive.

And for v1.05 take a look at the modified versions by Daniel Schulz

http://slothsoft.net/getResource.php/amber/patch/Ambermoon.v1.06.zip

I don't know much about v1.05 as I don't read German but I believe he's fixed most (all?) of the common bugs between v1.07 and v1.05. He has a list v1.05 bugs albeit not down to the specific bytes :

http://slothsoft.net/Ambermoon/EditorHelp/ (look at the section : MIR BEKANNTE BUGS IN AMBERMOON)

**Comment by Pyrdacor** [comment]
Slow is ok. I have to implement battles, too. I can read german so you can focus on the english version. Thanks for the links.

I also prefer a file by file approach. If you don't know in which file a bug may be located, feel free to ask. I am very familiar to the internal data structures and their relations already.

**Comment by a1exh** [comment]
Object_texts.amb sub-file 001.amb this file has a bug in it where it uses the word "GRANDSON" regardless of the gender of the main character. In my bugfix I change the text to "GRANDCHILD". HOWEVER I understand that the bugs like this surrounding gender are found in other places. Do you know if the scripting special markup (e.g. ~SELF~, ~LEAD~) has gender specific codes that I could look for?

**Comment by a1exh** [comment]
One bug was described as a typo "YELLOW CARD SPHER" changed to "YELLOW CARD SPHERE" but don't know what file this is in. I didn't see it in any *_texts.amb file. It is possibly an item? So in Save.00/Chest_data.amb ?

**Comment by Pyrdacor** [comment]
It is an item but they are stored in AM2_CPU and AM2_Blit.

Recently I found another issue. Don't know if this is fixed. When you leave the airship on forest moon you can trigger the witch broom message twice. We should change that by a map event condition.

**Comment by a1exh** [comment]
Thanks, that explains it, the patches that were made to AM2_CPU and AM2_BLIT by Hexaae in approx. 2001 (including 2-3 other fixes) I have unfortunately lost at sometime in the past. I think you discovered they were packed using imploder. I am able to depack using the Amiga tool xfddecrunch, the game still runs depacked, I can find the strings but I don't understand the data structure enough to insert bytes into strings without breaking something (e.g. string length).

**Comment by a1exh** [comment]
> Recently I found another issue. Don't know if this is fixed. When you leave the airship on forest moon you can trigger the witch broom message twice. We should change that by a map event condition.

No-one has reported this bug previously. A save game near by and technical description would be good.

**Comment by Pyrdacor** [comment]
The strings in AM2_CPU are a bit strange but luckily the item names have fix lengths. I can help editing them. You can decompress and compress the AM2_CPU / AM2_Blit with imploder-4 in WinUAE.

**Comment by Pyrdacor** [comment]
> > Recently I found another issue. Don't know if this is fixed. When you leave the airship on forest moon you can trigger the witch broom message twice. We should change that by a map event condition.
> 
> No-one has reported this bug previously. A save game near by and technical description would be good.

I can provide a savegame tomorrow. When you leave the airship on forest moon the first time with the witch's broom in inventory and then stumble upon a nearby tile an event is triggered with a text popup. The text says that a demon appears and crushes the broom. Then the broom is destroyed. But when you go to another tile nearby, the same popup is triggered again. This should not happen.

**Comment by a1exh** [comment]
v1.07 has an issue where if you're flying/sailing near the coordinates [0,0] (might be [1,1]) that there is a reference to a missing text block in 1Map_texts.amb that causes the Amiga version to crash. This was patched in two different ways, Metallic fixed it by adding a dummy text block 001.amb into 1Map_texts.amb so that it didn't crash. Meynaf said he removed the reference to the missing text block but he didn't say where or how he did this. The Meynaf patch doesn't modify 1Map_data.amb. I'd be curious to know what the underlaying cause of this is and how Meynaf patch works.

**Comment by a1exh** [comment]
One of the more aesthetic issues in v1.07 are the translations of the stone head limericks (rhymes). In German they all rhyme and are fun but whoever translated them did the equivalent of Google translate and didn't attempt to re-create the style of the German versions. Thinking up new text is relatively easy. Getting them into the files without making a mistake is harder. This is where the data-writers may help.

e.g. 2Map_texts.amb - 103.amb

The German version is this :
>DEINE STIMME KENN ICH NICHT,
DU KANNST HIER NICHTS ERREICHEN!
DOCH SAGST DU MIR DAS RICHTGE WORT,
DANN LASS ICH MICH ERWEICHEN.

>DICH DRÄNGT DER DURST?
DU SUCHST DEN WEIN?
WOHLAN DENN,
GEH DEN GANG HINEIN!

But the translation falls so flat
>I DO NOT KNOW YOUR VOICE,
YOU CANNOT ACHIEVE ANYTHING HERE!
BUT IF YOU SAY THE RIGHT THING,
THEN I SHALL LET YOU GO.

> "ARE YOU THIRSTY?
WOULD YOU LIKE SOME WINE?
PERHAPS
THEN GO INTO THE PASSAGEWAY!"

A German fan pointed out all these translation issues even suggesting some better translations :

>I DO NOT RECOGNISE YOUR VOICE,
WHICH LEAVES ME NO OTHER CHOICE!
I WILL ONLY LET YOU THROUGH,
IF YOU ANSWER ME THIS CLUE!

> YOU SUFFER REAL THIRST?
WOULD YOU LIKE SOME WINE?
LOOK IN THE CORNER FIRST,
DO NOT WASTE ANY TIME!

However modifications like this manually without any checking, especially with mark-up for bold etc. is why they have not been fixed yet.

**Comment by a1exh** [comment]
> When you leave the airship on forest moon the first time with the witch's broom in inventory and then stumble upon a nearby tile an event is triggered with a text popup. The text says that a demon appears and crushes the broom. Then the broom is destroyed.

This is interesting, I don't remember this. But Daniel Schulz reports for v1.05 that you can smuggle a broom onto Morag's moons using two spells (I can't read the German to know which spells they are).

> Hexenbesen nach Morag schmuggeln
Auf dem Wüstenmond wird ein eventuell mitgeführter Hexenbesen per Event zerstört. Über die Zauber "Wort des Markierens" und "Wort der Rückkehr" lässt sich das Triggern dieses Events umgehen, und damit der Besen als Transportmittel auf dem Mond benutzen.

**Comment by Pyrdacor** [comment]
The data writer will have no problem adding or editing map texts like the riddles.

About the broom event. I think the best solution would be to trigger the event right when leaving the airship.

**Comment by Pyrdacor** [comment]
Tomorrow I will give you 3 things:
- Confirmation about the Reg issue in 1.07
- Savegame for broom event on forest moon
- A small command line tool which can:
    - Export all map texts as txt files
    - Import the text files back into the game data

You can then edit all map texts to your liking.

**Comment by Pyrdacor** [comment]
Maybe better?

    You look for wine?
    Your throat is dry?
    The answer was fine.
    You can pass by.

**Comment by Pyrdacor** [comment]
@a1exh Here is the savegame on the forest moon. Go outside the airship. The text should popup. Then go 1 tile right or left. The text should popup again.
[Save.07.zip](https://github.com/Pyrdacor/Ambermoon.net/files/5487660/Save.07.zip)

**Comment by Pyrdacor** [comment]
@a1exh Confirmed. In english 1.07 Reg is still bugged. He is not marked as boss. You can cast Fear/Panic on him and he would flee with the items you actually need.

**Comment by a1exh** [comment]
Please can you confirm the technical information for the fix?

I extract sub-file 012.amb in monster_char_data.amb (the one which contains the string "REG HILL-GIANT") 

Using the information from here (thank you very much) : 

https://github.com/Pyrdacor/Ambermoon/blob/master/FileSpecs/Characters.md

Byte offset 0x0012 (Monster Flags) needs bit 2 (Boss) set to 1?

In v1.07 "Monster Flags" for REG HILL-GIANT is 0x02 and I change it to 0x06?

**Comment by Pyrdacor** [comment]
Reg has bit 1 set only for monster flags. It must be bit 2. He also has bit 0 of element set which might be a bug as only the guard demon has this bit set.

**Comment by Pyrdacor** [comment]
The tool to export and import texts is finished and is working on my end: https://github.com/Pyrdacor/Ambermoon/releases/download/v1.0.0/AmbermoonTools-Windows.zip

Usage:
- Best copy the file "AmbermoonTextImport.exe" on folder above the "Amberfiles" folder.
- Make a copy of the *.amb files you want to edit (the tool will also create backups, but for safety)
- Start a Windows CMD and navigate to the folder where the tool is
- Start the tool without parameters to see its usage

I used it like this to edit some texts in grandfather's house:
1. `AmbermoonTextImport.exe -e Amberfiles 2Map_texts.amb MyExportedTexts`
2. Edit the text files (in this case they are in "MyExportedTexts\2Map_texts.amb\258")
3. `AmbermoonTextImport.exe -i Amberfiles 2Map_texts.amb MyExportedTexts`

After restarting Ambermoon the texts should be updated. Let me know how it went. ;)

**Comment by a1exh** [comment]
> The sub-file should be 018.amb (sub-files are counted starting at 1).
I am using the Amiga Ambermoon tools by Metallic. They label the files in HEX so 018.amb is the same as (0x)012.amb.

All references to sub-files in my spreadsheet are in hex

**Comment by a1exh** [comment]
> If you don't know in which file a bug may be located, feel free to ask. I am very familiar to the internal data structures and their relations already.
In the bandits cellar, there are a few rooms where you have to push a button or pull a lever to open passages. One of those walls has a second, non-functional, button. It's similar to the stairs in Sansrie's temple and Luminor's Tower which have the wrong gfx and show up not down and vice versa. These are caused by an incorrect map entry (e.g. the wall should have only one button, and the wall tile with the second button should be a regular wall tile).

This is devilishly hard to get correct using a HEX editor. It would be good to be able to edit the MAP data and then test the dungeon to change mistakes when it comes to the MAP data writer.

**Comment by Pyrdacor** [comment]
> > The sub-file should be 018.amb (sub-files are counted starting at 1).
> 
> I am using the Amiga Ambermoon tools by Metallic. They label the files in HEX so 018.amb is the same as (0x)012.amb. All references to sub-files in my spreadsheet are in hex

Ah ok makes sense. Then this is ok.

**Comment by Pyrdacor** [comment]
> > If you don't know in which file a bug may be located, feel free to ask. I am very familiar to the internal data structures and their relations already.
> 
> In the bandits cellar, there are a few rooms where you have to push a button or pull a lever to open passages. One of those walls has a second, non-functional, button. It's similar to the stairs in Sansrie's temple and Luminor's Tower which have the wrong gfx and show up not down and vice versa. These are caused by an incorrect map entry (e.g. the wall should have only one button, and the wall tile with the second button should be a regular wall tile).
> 
> This is devilishly hard to get correct using a HEX editor. It would be good to be able to edit the MAP data in a more efficient manor or at least be able to test the dungeon to change mistakes when it comes to the MAP data writer.

I agree. But it is also hard to create a tool to edit such stuff in an easy way. Maybe you have some ideas. Command line tools can be created rather fast. I can also print all map properties by name (like "Tile[1,1]" or "MusicIndex", etc).

**Comment by Pyrdacor** [comment]
Did you try the text importer?

**Comment by a1exh** [comment]
ASCII representation of the dungeon levels using different characters for the tiles?

I have not yet tried the text importer. I will try it soon as I have several files I hand edited and changed string lengths and I want to see if they are the same.

**Comment by Pyrdacor** [comment]
What do you mean by ASCII representation of dungeon levels?

**Comment by a1exh** [comment]
Did you never play text based MUDs? (I am showing my age here). I was thinking about printing a 2D layout to the console window with different characters for wall textures etc.

Thinking about it , it is a pointless idea, the better idea would be to be able to edit the map data somehow, perhaps as text files, write them back to the data files, launch the game and interact.

Perhaps this could be done in conjunction with the Savegame writer? If you can quickly and easily place your party near to the thing you're editing?

**Comment by Pyrdacor** [comment]
Sorry I misread your comment. Of course I know what you mean. But wall and objects indices can be greater than 100. We will ran out of printable characters quickly.

Maybe I can export maps as xml or json.

**Comment by a1exh** [comment]
Another curiosity that maybe the data writers can answer. 2Map_texts.amb sub-file 1C6.amb and 1B7.amb are *ALMOST* the same file. The two files must exist this way in the German version because they have been translated separately and as a result each one contains slightly different English translations for different parts. I am curious why there are two sub-files almost the same?

**Comment by Pyrdacor** [comment]
Are you sure about that? I just extracted the files and they are not really similar.

454 (0x1C6) is HANGAR ON MORAG. It has 5 texts including the map name.
447 (0x1B6) is PALACE OF S'ENDAR. It has 12 texts including the map name.

The texts themselves are very different too. Where did you find this information?

**Comment by Pyrdacor** [comment]
You are right. The two files are identical in german. In english the translation is a bit different for 3 of the 5 texts. They were translated separately but the texts have the same meaning in english too.

**Comment by Pyrdacor** [comment]
Ok map 0x1C6 can be accessed from the airship on Morag and from the town of S'Angrila.
Map 0x1B7 can only be accessed from the airship on Morag.

There is a condition which checks for global variable bit 222. If you leave the airship and the variable bit is 0, map 0x1C6 is entered, otherwise map 0x1B7. I will check where variable 222 is set and how both maps differ.

**Comment by Pyrdacor** [comment]
Ok it looks like map 0x1B7 contains the NPC S'Riel who is nearly dead. When talking to him he will tell you about the plan of the Morag people and then teleports you to his tower. Then the global variable 222 is set to 0 so later when leaving the airship on Morag you will enter map 0x1C6.

Background: When you first enter morag via airship you will enter map 0x1B7. It contains an event which teleports you to the prison cell where S'Riel is dying. After talking to him he will teleport you to his tower. After that you will never enter map 0x1B7 again but map 0x1C6 instead which does not contain the teleport event to the prison.

The different texts are just for observing some things in the hangar and are slightly different.

**Comment by a1exh** [comment]
It is not just these two files, it is also 1C7.amb and 1B8.amb S'ANGRILA PRISON

Daniel Schulz has some notes about this, but I do not fully understand as they are not technical

> **HANGAR AUF MORAG und GEFÄNGNIS - S'ANGRILA sind nicht auf dem Wüstenmond**
Die Karten "HANGAR AUF MORAG" und "GEFÄNGNIS - S'ANGRILA" sind als "auf Lyramion" gekennzeichnet, sind aber eigentlich auf dem Wüstenmond. Dadurch lassen sich dort die Zaubersprüche WORT DES MARKIERENS und WORT DER RÜCKKEHR wirken, was den oben beschriebenen Trick ermöglicht. (Das gilt allerdings nur für das erstmalige Betreten des Mondes; nach dem Event mit dem sterbenden Moraner werden fortan 2 Duplikate dieser Karten verwendet, die korrekt als "auf Morag" gekennzeichnet sind.)

> **HANGAR ON MORAG and S'ANGRILA PRISON are not on the desert moon.** 
The map locations "HANGAR ON MORAG" and "S'ANGRILA PRISON" are marked as "on Lyramion", but are actually on the desert moon MORAG. This allows the WORD OF MARKING and WORD OF RETURN spells to work there, which enables the trick described above. (This only applies to the first time you step onto the moon; after the event with the dying Moraner, 2 duplicates of these map locations will be used, which are correctly marked as "on Morag".)

So global variable 222 must be this event "The dying Moraner" (I don't remember what this is in the game)

**Comment by Pyrdacor** [comment]
We can just change the world of these 2 maps. This will fix the bug. The world byte is documented in the other repo.

**Comment by a1exh** [comment]
Any chance you could add a switch to the data writers so they wrote filenames with HEX numbering. It would make my life a bit easier

**Comment by a1exh** [comment]
I tried the extract and insert to modify the file. It worked well. A few minor notes.

It has removed lots of $20 characters from within the sub-file text blocks. (Whitespace?) Is this deliberate? It does make RAW diff more difficult.

It didn't LOB pack the sub-files. Was that to be expected?

**Comment by a1exh** [comment]
More importantly the files can be loaded and extracted by the existing Amiga tools. All the fixes I am attempting, unfortunately I don't have save games anywhere near the affected areas so I cannot test in-game.

A switch to preserve whitespace would be good, so the files could be easily diffed against the originals.

**Comment by Pyrdacor** [comment]
I will add switches for hex names and whitespace preserving.

About the LOB compression. The text files use the AMNP format. This is an archive where all files are automatically JH encoded and optionally LOB compressed. As the LOB compression is optional I only use LOB if the file size would be smaller. So my tools will always produce the smallest file sizes possible. If you want, I could also add a switch for forcing LOB compression on AMNP but I don't know if this would be desirable.

**Comment by a1exh** [comment]
No that is fine, these were small files and to be honest I didn't check to see if the originals were LOB packed

**Comment by Pyrdacor** [comment]
Indeed the originals are. But I guess they just used LOB compression manually on those files without checking if it really is of any use on a file by file basis.

**Comment by Pyrdacor** [comment]
Which format for filenames do you prefer? Just 1C0.txt? Or $1C0.txt or 0x1C0.txt? Do you want the directory names (which represent the map index) also in hex?

**Comment by a1exh** [comment]
The important bit <for me> is the sub-file directory name. 1Map_texts.amb/**1C0**/<anything>.txt

**Comment by Pyrdacor** [comment]
Ok I added all the features but I will dig into the larger file size a bit before releasing.

**Comment by Pyrdacor** [comment]
The archive format usage is a bit are strange.

AMNP uses JH and then optionally compresses with LOB.
AMPC doesn't use JH but always LOB.

In the first case (which is used for the text files) the LOB compression isn't very good cause JH leaves very random byte data which is not compressable in a good way. So I don't understand why AMNP doesn't use the other way around: first LOB then JH. Moreover I don't get why the original data files were compressed very good.

If I use my own LOB algorithm and use AMPC (LOB only) the file size auf 2Map_texts.amb is even smaller than that of the original. But with AMNP it is much larger which makes sense as JH + later LOB is bad. I guess there is something about the AMNP format I didn't understand.

**Comment by Pyrdacor** [comment]
Ok I got it. AMNP is indeed first LOB compressed and then JH encoded. This makes much more sense. I will fix that. It worked cause the compression was always bad the other way around so he never compressed. :D

**Comment by Pyrdacor** [comment]
@a1exh I just released 1.0.2 of the tools. Can you try it please? https://github.com/Pyrdacor/Ambermoon/releases/download/v1.0.2/AmbermoonTools-Windows.zip

There are 3 new options:
--preserve-whitespace (or -p)
--preserve-zero-bytes (or -z)
--hex-subfile-names (or -h)

You can also use -pzh to use all 3 options in a shorter way.

Examples:

AmbermoonTextImport.exe -e Amberfiles 2Map_texts.amb Foo -pzh
AmbermoonTextImport.exe -i Amberfiles 2Map_texts.amb Foo -pzh

Note: Always use the same options for export and import.

The compression and AMNP writing should be fixed. The 2Map_texts.amb is even a bit smaller as in original.

**Comment by a1exh** [comment]
Thanks for the quick fixes. -h works great.

The -p option doesn't seem quite right. The extracted files now include what I thought was the end of textblock marker (0x00) and in some cases whitespace (0x20) after the end of textblock marker I thought belonged to the start of the next textblock?

I wanted to see if I can use the tools to fix bugs where there is a missing end-of-textblock marker. There are at least 3 of these in v1.07 which cause the text parser to either crash, not display anything or display garbage. 

2Map_texts.amb/105/003.txt 
3Map_texts.amb/153/010.txt
NPC_texts.amb/03A/007.txt

It has done a much better job than I thought it would do. I was expecting the textblock.txt file to contains two entries, and have to rename all the .txt files that come after by incrementing the number by one. But it has split them correctly. However with files containing non ASCII / UTF-8 characters now (e.g. 0x00) I don't think you can use a text editor to fix them.

**Comment by Pyrdacor** [comment]
The text files don't really use "end-of-text-markers" (I call them terminating zeros). Instead each text-file has a header which specifies the "total number of texts" and for each of them a "text length".

I original data some texts have spaces (0x20) or zero-bytes (0x00) at the end in any order (therefore you also see spaces behind 0-bytes). But they don't matter. This is why I removed them before as I don't need them in Ambermoon.net.

If you want "end-of-text-markers" I think you have to fix the "text length" entry at the beginning of the file. But it should be ok to just fix the text files with my tools. They will write the correct text length then.

**Comment by Pyrdacor** [comment]
So in short you should use the -p and -z options only for research. When editing the data just avoid these options. The text archive will be much smaller and nobody needs the spaces and zeros. Neither original game nor Ambermoon.net. :)

**Comment by a1exh** [comment]
Understood.

I was sure the Amiga code used both "text length" and "terminating zeros" and if they don't match, the text parser malfunctions. 

The examples above I believe have correct "text length" but wrong "terminating zeros" and they make the Amiga text parser malfuction

**Comment by Pyrdacor** [comment]
Hm ok I didn't check that. But my tools will also take care of it. If the text does not end with a terminating zero it will automatically add one and will also increase the text length.

**Comment by Pyrdacor** [comment]
To put it another way: when using the text importer it will automatically fix terminating zeros. You only have to correct the texts. I strongly suggest to avoid the new -p and -z options for export and import. Do this only for temporarily diff. Always edit the files you exported without these options and also import without them.

**Comment by a1exh** [comment]
I will. It just makes it harder to diff with the originals (for confirmation the tool is doing what it should) because byte offsets change.

**Comment by a1exh** [comment]
I would be nervous to extract and insert everything with this tool **today** and then release them for the Amiga untested because they are so different.

**Comment by Pyrdacor** [comment]
I understand. Can you provide an edited version and a list of fixes that it contains? Maybe someone here (including me) could test it then.

**Comment by Pyrdacor** [comment]
Or you could create an issue here where you publish beta releases. I guess a few fellas here might test it then. I myself would use it too in Ambermoon.net testings.

**Comment by a1exh** [comment]
It would be good (for me) if you could flag any difference between the text-block length in the sub-file header and the number of (valid) characters between "terminating zeros". An optional warning during extraction? Could that be added? I am sorry if this is monopolising your time. Feel free to make it low priority.

**Comment by a1exh** [comment]
> Or you could create an issue here where you publish beta releases. I guess a few fellas here might test it then. I myself would use it too in Ambermoon.net testings.

I can start making v1.07 patches using your tools very soon. I am happy that every reported bug (that was data related and not executable related) has been examined, understood and most now have hand made diff-able patches done using the old tools.

There are just four which are data related but are inside am2_cpu and am2_blit that I can't patch myself as I don't have enough knowledge of how to increase string length inside this file.

**Comment by Pyrdacor** [comment]
Glad to hear. About the warning. I also could add the possibility to diff two files (e.g. 2Map_texts.amb) and generate a human readable diff for each subfile inside a log file. If you want I can write it also into csv (loadable with excel) etc.

**Comment by Pyrdacor** [comment]
I looked a bit into 68k assembler. But I don't have the time to learn it. But if I'm right I would guess that strings from the data section are read by loading their starting address into an address register. There is an instruction called LEA (load effective address). It has a specific opcode. Moreover each instruction has some information how the operands of an instruction are addressed. Assuming they (or the compiler) used static offsets this leads us to immediate data access. I can't explain this in detail now but I will do some tests to find out if I'm right. If so we just need to find the LEA instruction which loads the string we want to change. And then we have to find all LEAs that load strings with a higher address. For all of them we have to increase their offset by the length change. This might be enough. Maybe we also need to adjust a static length value but I guess they will be near the LEA instruction. I have an idea how to easily find a LEA to a data address. So maybe I could indeed help editing texts in the assemblies. But I am not 100% sure. If you want we could try it with a specific text. Just tell me which file and text you want to edit. And of course in which game version/data.

**Comment by Pyrdacor** [comment]
@a1exh Did you read the diff stuff and AM2_CPU text edits?

**Comment by a1exh** [comment]
> I also could add the possibility to diff two files (e.g. 2Map_texts.amb) and generate a human readable diff for each subfile inside a log file. If you want I can write it also into csv (loadable with excel) etc.

The problem I see is that ordinary 1:1 byte diffs would give lots of differences between the original and those having been post processed using the writer. Just extracting and re-inserting changes them quite a bit.

> If you want we could try it with a specific text. Just tell me which file and text you want to edit. And of course in which game version/data.

v1.07

1. 'MOTHER-O-PEARL CHAI' change to 'MOTHER OF PEARL CHAIN'
2. 'YELLOW CARD SPHER' change to 'YELLOW CARD SPHERE'
3. 'PELANI'S KEY' change to 'PELANIS' KEY' (_this should be easy as string length stays the same_)

**Comment by Pyrdacor** [comment]
Item names have fixed text lengths of 18 (19 with terminating zero). This max length can not be changed as you would have to change the whole item data structure.

So you have to limit your texts to the 18 characters. But within you can change the text length to your liking as the rest of the 18 bytes are just filled with zeros and the data structure remains intact.

So changing item names is no problem at all unless you stay in the 18 character bounds.

We need no 68k for that. This would only be necessary for things like game messages etc as the have no fixed structure.

**Comment by Pyrdacor** [comment]
> > I also could add the possibility to diff two files (e.g. 2Map_texts.amb) and generate a human readable diff for each subfile inside a log file. If you want I can write it also into csv (loadable with excel) etc.
> 
> The problem I see is that ordinary 1:1 byte diffs would give lots of differences between the original and those having been post processed using the writer. Just extracting and re-inserting changes them quite a bit.
> 

But if you use options `-pz` does the re-inserting still change the bytes?

**Comment by Pyrdacor** [comment]
I would strongly recommend **not** to change the max item name length. This is a very radical change to the game data. It is like changing the structure of the savegame data. Don't do it. Then better think of alternative names that fit into 19 characters or use abbreviations.

**Comment by Pyrdacor** [comment]
@a1exh Hi Alex. How is it going with patching? Any progress with the tools? Do you have a plan for a new patch release?

**Comment by a1exh** [comment]
I'm a single parent of a 5yo and in the UK the schools are closed until March so I've got no free time until then but I can do an hour or two here and there.

I'm keeping the v1.07 buglist upto date and slowly producing all the "diff" files.

https://docs.google.com/spreadsheets/d/1as5W8gibm-MTb9VEqpkfgtwWviqjQx96A3NmcvzX98A/edit?usp=sharing

**Comment by Pyrdacor** [comment]
> We can just change the world of these 2 maps. This will fix the bug. The world byte is documented in the other repo.
> 
> But I wonder why they used 2 maps. Events can be deactivated. This is often used. Don't know why 2 maps are necessary.

@a1exh I just watched a let's play and now I know why these maps are there twice and why two of them are marked as "on Lyramion". This was done on purpose.

The indoor player graphic has 3 versions. One for each world. The one for Morag has the head of a lizard.

The reason for the lizard head is that S'Riel gives you this as a disguise in the prison. But when you enter Morag for the first time you haven't talked to him. So you don't have the disguise yet. That's why you must look like a normal person. So they used the trick to mark the map as "on Lyramion" on these two maps (hangar and prison). Then when you have the disguise, they replaced them with two other maps that are marked as "on Morag" and you look like a lizard. It's a bit odd that they've done it that way but I guess the data designed did not allowed it in another way since the appearance is based on the map's world.

Now at least it makes sense that there are two duplicates maps and that they use the wrong world. But they didn't thought about "Word of return". :)
